### PR TITLE
Add gazebo reach env

### PR DESCRIPTION
**Pull request by hjzh4**

### CODE REVIEW

File: contrib/ros/worlds/empty_world/world.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import collections
```
Comment (ryanjulian) [inline contrib/ros/worlds/empty_world/world.py]: There's no need to create a whole package subfolder for a single module.

Just put it in `worlds/empty_world.py`

File: contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import rospy
+
+from contrib.ros.envs.sawyer.reach_env import ReachEnv
+from garage.algos.trpo import TRPO
+from garage.baselines.linear_feature_baseline import LinearFeatureBaseline
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py]: please use shortened imports

File: contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import rospy
+
+from contrib.ros.envs.sawyer.reach_env import ReachEnv
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py]: shortened imports

Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py]: this environment is usually called `ReacherEnv` in the literature

File: contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+from garage.algos.trpo import TRPO
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py]: shortened imports

File: contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+from garage.baselines.linear_feature_baseline import LinearFeatureBaseline
+from garage.envs.util import spec
+from garage.misc.instrument import run_experiment
+from garage.policies.gaussian_mlp_policy import GaussianMLPPolicy
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py]: shortened imports

File: contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    reach_env = ReachEnv(
+        initial_goal,
+        initial_joint_pos=INITIAL_REAL_ROBOT_JOINT_POS,
+        simulated=False)
+
+    rospy.on_shutdown(reach_env.shutdown)
+
+    reach_env.initialize()
+
+    env = reach_env
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_real_sawyer_reach.py]: unnecessary variable. just do `env = ReacherEnv(...)` above.

File: contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    reach_env = ReachEnv(
+        initial_goal,
+        initial_joint_pos=INITIAL_SIM_ROBOT_JOINT_POS,
+        simulated=True)
+
+    rospy.on_shutdown(reach_env.shutdown)
+
+    reach_env.initialize()
+
+    env = reach_env
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py]: unnecessary variable.

File: contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import rospy
+
+from contrib.ros.envs.sawyer.reach_env import ReachEnv
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py]: shorten imports

File: contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+from garage.algos.trpo import TRPO
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py]: shorten imports

File: contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+from garage.baselines.linear_feature_baseline import LinearFeatureBaseline
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py]: shorten imports

File: contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+from garage.envs.util import spec
+from garage.misc.instrument import run_experiment
+from garage.policies.gaussian_mlp_policy import GaussianMLPPolicy
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/launchers/trpo_sim_sawyer_reach.py]: shorten imports

File: contrib/ros/envs/sawyer/push_env.py [hunk 12]
```diff
@@ -8,7 +8,7 @@
 
 from contrib.ros.envs.sawyer.sawyer_env import SawyerEnv
 from contrib.ros.robots.sawyer import Sawyer
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/push_env.py]: shorten imports

File: contrib/ros/envs/sawyer/push_env.py [hunk 13]
```diff
@@ -8,7 +8,7 @@
 
 from contrib.ros.envs.sawyer.sawyer_env import SawyerEnv
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/push_env.py]: shorten imports

File: contrib/ros/envs/sawyer/push_env.py [hunk 14]
```diff
@@ -8,7 +8,7 @@
...
 from contrib.ros.robots.sawyer import Sawyer
-from contrib.ros.worlds.block_world.world import BlockWorld
+from contrib.ros.worlds.block_world import BlockWorld
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/push_env.py]: shorten imports

File: contrib/ros/envs/sawyer/reach_env.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
+"""
+Reach task for the sawyer robot
+"""
+import collections
+
+import gym
+import numpy as np
+
+from contrib.ros.envs.sawyer.sawyer_env import SawyerEnv
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/reach_env.py]: shorten imports

File: contrib/ros/envs/sawyer/reach_env.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+from contrib.ros.robots.sawyer import Sawyer
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/reach_env.py]: shorten imports

File: contrib/ros/envs/sawyer/reach_env.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+from contrib.ros.worlds.empty_world import EmptyWorld
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/reach_env.py]: shorten imports

File: contrib/ros/envs/sawyer/reach_env.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+from garage.core import Serializable
+
+
+class ReachEnv(SawyerEnv, Serializable):
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/reach_env.py]: this environment is usually called "Reacher" in the literature.

File: contrib/ros/worlds/empty_world.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
+import collections
+
+import gym
+import numpy as np
+
+from contrib.ros.worlds.world import World
```
Comment (ryanjulian) [inline contrib/ros/worlds/empty_world.py]: shorten imports

File: contrib/ros/envs/sawyer/reach_env.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+        self.initial_goal = initial_goal
+        self.goal = self.initial_goal.copy()
+
+        self._sawyer = Sawyer(
+            initial_joint_pos=initial_joint_pos,
+            control_mode=robot_control_mode)
+        self._world = EmptyWorld(simulated)
+
+        SawyerEnv.__init__(
+            self, simulated=simulated, robot=self._sawyer, world=self._world)
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/reach_env.py]: doesn't SawyerEnv set `self._robot` and `self._world`?

File: contrib/ros/envs/sawyer/reach_env.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        """
+        Returns a Space object
+        """
+        return gym.spaces.Box(
+            -np.inf,
+            np.inf,
+            shape=self.get_observation().observation.shape,
+            dtype=np.float32)
+
+    def sample_goal(self):
```
Comment (ryanjulian) [inline contrib/ros/envs/sawyer/reach_env.py]: please use the annotation `@overrides` when overriding a superclass method which is non-empty

```python
from rllab.misc.overrides import overrides


class Foo:

    def bar(self):
        print("bar")


class Baz(Foo):
    
    @overrides
    def bar(self):
        print("buzz")
```

Comment (hjzh4) [reply contrib/ros/envs/sawyer/reach_env.py]: Good point. But isn't the superclass method for this method 'sample_goal' is empty?

File: contrib/ros/envs/sawyer/reach_env.py (hunk 20 above)
Comment (hjzh4) [reply contrib/ros/envs/sawyer/reach_env.py]: oh right!

**Pull request merged**